### ISSUE TITLE
Implement data conversion applet

### ISSUE THREAD

**Issue by Tomaz-Vieira** [ilastik/webilastik#13]
The ebrains Image Service is very cumbersome to users and also has a few bugs. We should implement an applet to perform data conversion as part of the pixel classification workflow.

**Comment by PolarBean** [comment]
1- ) For sure converting to NG chunks should be easy, they're pretty similar to DZI just in 3D
2- ) It could produce non zipped files, but it was designed around making transfers to the data proxy easier (transferring chunks can add hours to the export time). We also do not call the entire file into our viewers as this would defeat the purpose of pyramid viewers. Instead what we do is make range requests to the volume for particular chunks, this is a best of both worlds approach and relatively straight forward since zip volumes are appended with an index that indicates the starting bit and size of each file in the archive. We have already implemented this logic https://github.com/Tevemadar/NetUnzip and I also have an implementation in node js if you're interested. But we are capable of outputting unzipped NG chunks, its just a matter of how comfortable you are with extremely slow export times. 
3- ) This is a good question, we currently only use an openshift vm to ingest images, I have tried with multiple users and it handles it easily. We could also enable scaling to more VM's if there is high demand. I wonder how much of an advantage running this operation on a HPC session would be (ie; I imagine you only allocate a CPU to the job) but we can of course explore both options.

**Comment by Tomaz-Vieira** [comment]
1-) great =D

2-) I looked at the code there. That's really a great idea, to just read records straight out of the zip file. Some questions on that:

2.1-) That does rely on the web server that is hosting the .zip to honor your request for specific byte-offsets of the file, right? I guess it's mostly not an issue for us as long as the Swift Object Storage can do that, which I presume is the case;

2.2-) I would need to convince Neuroglancer to make its tile requests using this read-from-zip-archive style, rather than expecting the tiles to be accessible via their file names. Even if I tried some redirect shenanigans on my webserver it  wouldn't be enough because the image tiles inside the zip would be compressed, and Neuroglancer isn't expecting that. This is all doable, but a bit more time intensive.

2.2.1-) Actually, come to think of it, you could just store the raw tiles in the `.zip` without compressing them, right? This would mean that just some redirect trickery would be enough to get neuroglancer to access the tiles inside the zip

3-) There are some niceties from the HPC, I'm not sure if all would be relevant for us:
 - A node has a lot of memory, so if you need to load an entire multi-GB image to memory before converting, then you can be pretty sure that it will fit;
 - Compression can be pretty compute-intensive, and depending on the compression scheme it could be done in parallel, so it's nice to have the 128 cores of the HPC working on that =)
 - By using an HPC session (that is exclusive to each user) we can't overwhelm the web server; worst case scenario we just crash that user's session.
 - Of course, if comes with the annoyance of having to allocate an HPC session. For webilastik users that's ok, since they need one anyway, but might not be ideal for other tools.

But I guess we can also worry about scaling once things start to break

**Comment by Tomaz-Vieira** [comment]
Cool =)
Is your tool ready (enough) for production? If so I can just link users to it if they try to open something like a `tiff` or a `png`.

Anyway, I'll halt development of the conversion tool on my side since you're further along and instead I'll start working on the other end;  making neuroglancer read out of the zip files. I'll also try to draw as much inspiration out of your implementation as I can =)

**Comment by PolarBean** [comment]
yeah our tool is very close to production ready :) I think that is the best approach, if you focus on implementing support for zipped ng chunks we can focus on the ingestion. As well I would say it would be good if you could handle zipped DZI as well since that would save having to double ingest 2D data formats but we can continue that conversation on rocket chat
